cmd/query/app/server.go

```diff
@@ -297,52 +295,58 @@ func (s *Server) Start() error {
 		}
 		if err != nil && !errors.Is(err, http.ErrServerClosed) && !errors.Is(err, cmux.ErrListenerClosed) && !errors.Is(err, cmux.ErrServerClosed) {
-			s.logger.Error("Could not start HTTP server", zap.Error(err))
+			s.Logger.Error("Could not start HTTP server", zap.Error(err))
+			s.ReportStatus(component.NewFatalErrorEvent(err))
+			return
 		}
-		s.logger.Info("HTTP server stopped", zap.Int("port", httpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
-		s.healthCheck.Set(healthcheck.Unavailable)
-		s.bgFinished.Done()
+		s.Logger.Info("HTTP server stopped", zap.Int("port", httpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
+		s.ReportStatus(component.NewStatusEvent(component.StatusStopped))
+		
 	}()
 
 	// Start GRPC server concurrently
 	s.bgFinished.Add(1)
 	go func() {
-		s.logger.Info("Starting GRPC server", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
+		defer s.bgFinished.Done()
+		s.Logger.Info("Starting GRPC server", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
 
 		err := s.grpcServer.Serve(s.grpcConn)
 		if err != nil && !errors.Is(err, cmux.ErrListenerClosed) && !errors.Is(err, cmux.ErrServerClosed) {
-			s.logger.Error("Could not start GRPC server", zap.Error(err))
+			s.Logger.Error("Could not start GRPC server", zap.Error(err))
+			s.ReportStatus(component.NewFatalErrorEvent(err))
+			return 
 		}
-		s.logger.Info("GRPC server stopped", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
-		s.healthCheck.Set(healthcheck.Unavailable)
-		s.bgFinished.Done()
+		s.Logger.Info("GRPC server stopped", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
+		s.ReportStatus(component.NewStatusEvent(component.StatusStopped))
 	}()
 
 	// Start cmux server concurrently.
 	if !s.separatePorts {
 		s.bgFinished.Add(1)
 		go func() {
-			s.logger.Info("Starting CMUX server", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
+			defer s.bgFinished.Done()
+			s.Logger.Info("Starting CMUX server", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
 
 			err := cmuxServer.Serve()
 			// TODO: find a way to avoid string comparison. Even though cmux has ErrServerClosed, it's not returned here.
 			if err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
-				s.logger.Error("Could not start multiplexed server", zap.Error(err))
+				s.Logger.Error("Could not start multiplexed server", zap.Error(err))
+				s.ReportStatus(component.NewFatalErrorEvent(err))
+				return
 			}
-			s.logger.Info("CMUX server stopped", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
-			s.healthCheck.Set(healthcheck.Unavailable)
-			s.bgFinished.Done()
+			s.Logger.Info("CMUX server stopped", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
+			s.ReportStatus(component.NewStatusEvent(component.StatusStopped))
```

Comment:
I was wrong, we don't need this one as it's done at the end of the Close function.

Reply:
Yes, I had added it separately in the Close function itself.